Splits combine_remediations.py into 2 scripts

jan-cerny opened the pull request:
#### Description:
Splits combine_remediations.py into 2 scripts.

First script (combine_remediations.py) gathers all remediation snippets, resolves Jinja macros and outputs all of the snippets to a single directory, currently set to CMake as `build/<product>/all_fixes/<language>`.

Second script (generate_fixes_xml.py) merges all the remediation snippets from this directory into a single XML file `<language>-fixes.xml`.  The scripts are called by CMake during the build and generate_fixes_xml.py depends on outputs of combine_remediations.py.

#### Rationale:
We will use this when we will generate Ansible playbook for each rule for each profile.

pep8speaks commented:
Hello @jan-cerny! Thanks for submitting the PR.

 - In the file [`build-scripts/combine_remediations.py`](https://github.com/ComplianceAsCode/content/blob/9f3910cce84f1b49a1853a257288c3440203069f/build-scripts/combine_remediations.py), following are the PEP8 issues :

> [Line 35:13](https://github.com/ComplianceAsCode/content/blob/9f3910cce84f1b49a1853a257288c3440203069f/build-scripts/combine_remediations.py#L35): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

 - In the file [`build-scripts/generate_fixes_xml.py`](https://github.com/ComplianceAsCode/content/blob/9f3910cce84f1b49a1853a257288c3440203069f/build-scripts/generate_fixes_xml.py), following are the PEP8 issues :

> [Line 62:29](https://github.com/ComplianceAsCode/content/blob/9f3910cce84f1b49a1853a257288c3440203069f/build-scripts/generate_fixes_xml.py#L62): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/1db38f67-22db-4bcf-abd8-9bb314baf23d) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/1db38f67-22db-4bcf-abd8-9bb314baf23d) completed: **2 new issues, 5 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/a470f2db-adba-4225-bd44-983ea191ab4a) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/a470f2db-adba-4225-bd44-983ea191ab4a) completed: **2 new issues, 5 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/3006e8b9-69c8-46f5-ac73-197b1e024900) was created.

redhatrises commented on cmake/SSGCommon.cmake:
```diff
@@ -238,16 +238,29 @@ macro(_ssg_build_remediations_for_language PRODUCT LANGUAGES)
     foreach(LANGUAGE ${LANGUAGES})
       file(GLOB EXTRA_LANGUAGE_DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/fixes/${LANGUAGE}/*")
       file(GLOB EXTRA_SHARED_LANGUAGE_DEPENDS "${SSG_SHARED}/fixes/${LANGUAGE}/*")
+      set(ALL_FIXES_DIR "${CMAKE_CURRENT_BINARY_DIR}/all_fixes/${LANGUAGE}")
```
Can we not just put all remediation scripts in the `fixes` directory rather than just having `all_fixes` and `fixes`?

jan-cerny replied:
Yes, I agree, I also don't like it.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/2695af92-48d7-4586-ac2f-9f379a17f9ba) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/2695af92-48d7-4586-ac2f-9f379a17f9ba) completed: **2 new issues, 8 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/f9a68a9e-ea90-4d4e-8257-83bd16159615) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/f9a68a9e-ea90-4d4e-8257-83bd16159615) completed: **8 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/83107ec7-4b21-4b09-ae97-81dbd843a508) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/83107ec7-4b21-4b09-ae97-81dbd843a508) completed: **95 Issues, 0 Patches**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/be55c9d4-cf1c-472c-9a43-a11c2414e7e4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/be55c9d4-cf1c-472c-9a43-a11c2414e7e4) completed: **97 Issues, 0 Patches**

jan-cerny commented:
@redhatrises 

> Can we not just put all remediation scripts in the `fixes` directory rather than just having `all_fixes` and `fixes`?

It's more complicated, we need a cleanup. See https://github.com/ComplianceAsCode/content/issues/3816 for more details. 

I have created a temporary solution. First, the `generate_from_templates.py` will create a directory `fixes_from_templates`. Then, `combine_remediations.py` will create `fixes` directory, which will contain all the fixes (both explicit and generated).

matejak commented on ssg/build_remediations.py:
```diff
@@ -166,22 +166,10 @@ def get_populate_replacement(remediation_type, text):
-    """
-
+def _parse_remediation(fix_file_lines):
     mod_file = []
```
`mod_file` is a bad name. What comes to my first is that `remediation_content` is better.

matejak commented on ssg/build_remediations.py:
```diff
@@ -166,22 +166,10 @@ def get_populate_replacement(remediation_type, text):
-    update ssg.fixes.parse_platform(...).
-    """
-
+def _parse_remediation(fix_file_lines):
```
This function splits remediation content and metadata, which are parsed into a dictionary. Therefore, I would suggest to be more concrete and consider naming e.g. `split_remediation_content_and_metadata` or something like that.

matejak commented on ssg/build_remediations.py:
```diff
@@ -204,6 +192,32 @@ def parse_from_file(file_path, env_yaml):
     return remediation(mod_file, config)
 
 
+def parse_from_file(file_path, env_yaml):
```
Obviously, the naming suggestion apply here as well.

A Remediation class would solve the situation and avoid too long names, as it would read files either with or without jinja expansion, and it would have methods for metadata extraction.

matejak commented on ssg/build_remediations.py:
```diff
@@ -259,6 +273,26 @@ def write_fixes(remediation_type, build_dir, output_path, fixes):
+
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+    for fix_name in fixes:
```
You can iterate over `.items()` and have `for fix_name, fix in fixes.items(): ...`, allowing you to avoid using `fix_contents` and `config`.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/9eac475a-4b20-4f74-9463-0b5d12b5be1d) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/9eac475a-4b20-4f74-9463-0b5d12b5be1d) completed: **95 Issues, 0 Patches**

jan-cerny commented:
@matejak I have renamed variable `mod_file` to `remediation_contents`, renamed function `_parse_remediation` to `_split_remediation_content_and_metadata`, renamed function `parse_from_file` to `parse_from_file_with_jinja` and I have iterated dictionary intelligently.

matejak commented:
Oh, I thought that I have merged it already. Great job anyway!
